src/Symfony/Component/Security/Http/Authentication/AuthenticatorManager.php

```diff
@@ -0,0 +1,241 @@
+            }
+
+            // check the passport (e.g. password checking)
+            $event = new VerifyAuthenticatorCredentialsEvent($authenticator, $passport);
+            $this->eventDispatcher->dispatch($event);
+
+            // check if all badges are resolved
+            $passport->checkIfCompletelyResolved();
+
+            // create the authenticated token
+            $authenticatedToken = $authenticator->createAuthenticatedToken($passport, $this->providerKey);
+            if (true === $this->eraseCredentials) {
+                $authenticatedToken->eraseCredentials();
+            }
+
+            if (null !== $this->eventDispatcher) {
+                $this->eventDispatcher->dispatch(new AuthenticationSuccessEvent($authenticatedToken), AuthenticationEvents::AUTHENTICATION_SUCCESS);
+            }
+
+            if (null !== $this->logger) {
+                $this->logger->info('Authenticator successful!', ['token' => $authenticatedToken, 'authenticator' => \get_class($authenticator)]);
+            }
+
+            // success! (sets the token on the token storage, etc)
+            $response = $this->handleAuthenticationSuccess($authenticatedToken, $passport, $request, $authenticator);
+            if ($response instanceof Response) {
+                return $response;
+            }
+
+            if (null !== $this->logger) {
+                $this->logger->debug('Authenticator set no success response: request continues.', ['authenticator' => \get_class($authenticator)]);
+            }
+
+            return null;
+        } catch (AuthenticationException $e) {
+            // oh no! Authentication failed!
+            $response = $this->handleAuthenticationFailure($e, $request, $authenticator);
+            if ($response instanceof Response) {
+                return $response;
+            }
+
+            return null;
+        }
+    }
+
+    private function authenticateViaAuthenticator(AuthenticatorInterface $authenticator, Request $request): TokenInterface
+    {
+        return $authenticatedToken;
+    }
+
+    private function handleAuthenticationSuccess(TokenInterface $authenticatedToken, ?PassportInterface $passport, Request $request, AuthenticatorInterface $authenticator): ?Response
+    {
+        $this->tokenStorage->setToken($authenticatedToken);
+
+        $response = $authenticator->onAuthenticationSuccess($request, $authenticatedToken, $this->providerKey);
+        if ($authenticator instanceof InteractiveAuthenticatorInterface && $authenticator->isInteractive()) {
+            $loginEvent = new InteractiveLoginEvent($request, $authenticatedToken);
+            $this->eventDispatcher->dispatch($loginEvent, SecurityEvents::INTERACTIVE_LOGIN);
+        }
+
+        $this->eventDispatcher->dispatch($loginSuccessEvent = new LoginSuccessEvent($authenticator, $passport, $authenticatedToken, $request, $response, $this->providerKey));
```

Comment:
I'm a little worried about 2 things here:

1) This is dispatched even for `AnonymousAuthenticator`. On a technical low-level, sure, that makes sense. From a practical perspective, that makes no sense and will trip people up. Perhaps this event should only be dispatched when there is a `UserInterface` object in the token? Or can you think of a better way to make this not dispatched in this case? I think it will be important from a practical standpoint.

2) Kind of related, the `LoginSuccessEvent` has no way to get the `User` out - was pass the `Passport`, but that's a `PassportInterface` - it has no accessor for the User. We pass the token, but we all know how annoying it is to need to get the user from the token. Even if we stop dispatching this in the anonymous case (which would at least prevent `$token->getUser()` from returning a string), a `getUser(): UserInterface` would probably be helpful.